plugins/h3/h3.cpp

```diff
@@ -0,0 +1,247 @@
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#include "platform.h"
+#include "eclrtl.hpp"
+#include "h3.hpp"
+
+#include <vector>
+
+#define H3_VERSION "h3 plugin 1.0.0"
+ECL_H3_API bool getECLPluginDefinition(ECLPluginDefinitionBlock *pb)
+{
+    /*  Warning:    This function may be called without the plugin being loaded fully.
+     *              It should not make any library calls or assume that dependent modules
+     *              have been loaded or that it has been initialised.
+     *
+     *              Specifically:  "The system does not call DllMain for process and thread
+     *              initialization and termination.  Also, the system does not load
+     *              additional executable modules that are referenced by the specified module."
+     */
+
+    if (pb->size != sizeof(ECLPluginDefinitionBlock))
+        return false;
+
+    pb->magicVersion = PLUGIN_VERSION;
+    pb->version = H3_VERSION;
+    pb->moduleName = "lib_h3";
+    pb->ECL = NULL;
+    pb->flags = PLUGIN_IMPLICIT_MODULE;
+    pb->description = "ECL plugin library for uber h3\n";
+    return true;
+}
+
+#include <h3api.h>
+
+void toSetOf(bool &__isAllResult, size32_t &__lenResult, void *&__result, const H3Index *buff, int maxBuffLen)
+{
+    std::vector<H3Index> out;
+    for (int i = 0; i < maxBuffLen; ++i)
+    {
+        if (buff[i] != 0)
+            out.push_back(buff[i]);
+    }
+    __isAllResult = false;
+    __lenResult = out.size() * sizeof(H3Index);
+    __result = rtlMalloc(__lenResult);
+    memcpy(__result, &out[0], __lenResult);
+}
+
+namespace h3
+{
+
+//--------------------------------------------------------------------------------
+//                           ECL SERVICE ENTRYPOINTS
+//--------------------------------------------------------------------------------
+
+//  Indexing Functions ---
+ECL_H3_API unsigned __int64 ECL_H3_CALL index(ICodeContext *_ctx, double lat, double lon, uint32_t resolution)
+{
+    GeoCoord location;
+    location.lat = ::degsToRads(lat);
+    location.lon = ::degsToRads(lon);
+    return ::geoToH3(&location, resolution);
+}
+
+ECL_H3_API void ECL_H3_CALL center(ICodeContext *_ctx, size32_t &__lenResult, void *&__result, unsigned __int64 index)
+{
+    GeoCoord location;
+    ::h3ToGeo(index, &location);
+    __lenResult = 2 * sizeof(double);
+    __result = rtlMalloc(__lenResult);
+    double *cur = static_cast<double *>(__result);
+    *cur = ::radsToDegs(location.lat);
```

Comment:
Slightly weird style -I would have said *cur++ on the first line and *cur on the second